File: backend/app/services/agent_tool_domains/tasks.py
"""Task management domain — create/update/delete tasks with DB sync."""

import asyncio
import logging
import uuid
from datetime import datetime, timezone
from pathlib import Path

from sqlalchemy import select

from app.database import async_session
from app.models.task import Task

logger = logging.getLogger(__name__)


async def _manage_tasks(
    agent_id: uuid.UUID,
    user_id: uuid.UUID,
    ws: Path,
    args: dict,
) -> str:
    """Create / update / delete tasks in DB and sync to workspace."""
    from app.models.task import TaskLog
    from app.tools.workspace import _sync_tasks_to_file

    action = args["action"]
    title = args["title"]

    async with async_session() as db:
        if action == "create":
            task_type = args.get("task_type", "todo")
            task = Task(
                agent_id=agent_id,
                title=title,
                description=args.get("description"),
                type=task_type,
                priority=args.get("priority", "medium"),
                created_by=user_id,
                status="pending",
                supervision_target_name=args.get("supervision_target_name"),
                supervision_channel=args.get("supervision_channel", "feishu"),
                remind_schedule=args.get("remind_schedule"),
            )
            db.add(task)
            await db.commit()
            await db.refresh(task)

            if task_type == "todo":
                # Trigger auto-execution for todo tasks
                from app.services.task_executor import execute_task
                asyncio.create_task(execute_task(task.id, agent_id))
                await _sync_tasks_to_file(agent_id, ws)
                return f"✅ Task created: {title} — auto-execution started"
            else:
                # Supervision task — reminder engine will pick it up
                target = args.get('supervision_target_name', 'someone')
                schedule = args.get('remind_schedule', 'not set')
                await _sync_tasks_to_file(agent_id, ws)
                return f"✅ Supervision task created: '{title}' — will remind {target} on schedule ({schedule})"

        elif action == "update_status":
            result = await db.execute(
                select(Task).where(Task.agent_id == agent_id, Task.title.ilike(f"%{title}%"))
            )
            task = result.scalars().first()
            if not task:
                return f"No task found matching '{title}'"
            old = task.status
            task.status = args["status"]
            if args["status"] == "done":
                task.completed_at = datetime.now(timezone.utc)
            await db.commit()
            await _sync_tasks_to_file(agent_id, ws)
            return f"✅ Updated '{task.title}' from {old} to {args['status']}"

        elif action == "delete":
            from sqlalchemy import delete as sa_delete
            result = await db.execute(
                select(Task).where(Task.agent_id == agent_id, Task.title.ilike(f"%{title}%"))
            )
            task = result.scalars().first()
            if not task:
                return f"No task found matching '{title}'"
            task_title = task.title
            await db.execute(sa_delete(TaskLog).where(TaskLog.task_id == task.id))
            await db.delete(task)
            await db.commit()
            await _sync_tasks_to_file(agent_id, ws)
            return f"✅ Task deleted: {task_title}"

        return f"Unknown action: {action}"

File: frontend/src/api/domains/tasks.ts
/**
 * Tasks domain adapter — agent task management.
 */

import { get, post, patch } from '../core';
import type { Task } from '../../types';

export interface TaskCreateParams {
  title: string;
  description?: string;
  type?: 'todo' | 'supervision';
  priority?: 'low' | 'medium' | 'high' | 'urgent';
  assignee?: string;
  due_date?: string;
}

export interface TaskUpdateParams {
  title?: string;
  description?: string;
  status?: 'pending' | 'doing' | 'done';
  priority?: 'low' | 'medium' | 'high' | 'urgent';
}

export interface TaskLog {
  id: string;
  task_id: string;
  content: string;
  created_at: string;
}

export const taskApi = {
  list: (agentId: string) => get<Task[]>(`/agents/${agentId}/tasks/`),
  create: (agentId: string, data: TaskCreateParams) => post<Task>(`/agents/${agentId}/tasks/`, data),
  update: (agentId: string, taskId: string, data: TaskUpdateParams) =>
    patch<Task>(`/agents/${agentId}/tasks/${taskId}`, data),
  getLogs: (agentId: string, taskId: string) => get<TaskLog[]>(`/agents/${agentId}/tasks/${taskId}/logs`),
  trigger: (agentId: string, taskId: string) => post<void>(`/agents/${agentId}/tasks/${taskId}/trigger`),
};
